blog/_posts/2018-11-23-ml-compiler.md

```diff
@@ -0,0 +1,69 @@
+
+This enables some powerful tricks, even if you never write CUDA code yourself. For example, we can transparently fuse a large broadcast expression like `1 / (1 + exp(-x))`, *and* its backwards pass, into a single GPU kernel, and getting [significant speedups](https://arxiv.org/abs/1810.08297).
+
+## Taking Gradients
+
+Pushing the limits of reverse-mode differentiation, we have also come to see this as a [language-level problem](https://arxiv.org/abs/1810.07951). Differentiation is a symbolic transformation, which is the domain of compilers. Existing frameworks achieve this by *tracing* (effectively a form of *partial evaluation* or *abstract interpretation*). A new tensor type is introduced which records all the basic mathematical operations performed, yielding a graph (or symbolic expression) with the control flow and data structures of the host language removed. However, this presents a difficult tradeoff: we either accept the overhead of an interpreter (eager execution) or freeze user control flow and limit the kinds of models that can be built (static graphs).
+
+What if, instead, the “graph” were simply Julia’s own syntax? Taking this idea to its limit, we have built [Zygote](https://github.com/FluxML/Zygote.jl), which works directly on SSA-form IR and supports language features like control flow, recursion, data structures and macros. We can then put the generated SSA-form adjoint code through a compiler such as [LLVM](http://llvm.org/), and get all the benefits of traditional compiler optimization applied to both our forward and backwards passes. In addition, this approach opens the opportunity to extend that compiler infrastructure with more advanced and domain-specific optimizations, such as kernel fusion and compilation to accelerators such as TPUs. Similar approaches are being explored by the [Swift for TensorFlow](https://gist.github.com/rxwei/30ba75ce092ab3b0dce4bde1fc2c9f1d) and [Myia](https://arxiv.org/abs/1810.11530) developers in a renaissance of source-to-source AD techniques.
+
+A key advantage of Julia for this task is that it can be used to implement fundamental numerical libraries, like [differential equations solvers](http://juliadiffeq.org/) or [optimisation libraries](https://github.com/JuliaOpt/JuMP.jl); this neatly solves a growing need in the ML community, in which researchers backpropagate through high-performance code such as [ray tracers](https://people.csail.mit.edu/tzumao/diffrt/) and [physics engines](https://arxiv.org/abs/1611.01652), but the gradients must still be implemented by hand in C++. In contrast, since Julia’s implementations are written in Julia, everything from [ODEs](https://github.com/FluxML/model-zoo/blob/a243e8b192236c30064fcdb7a36f17f3b6823c34/other/diffeq/diffeq.jl) to [financial princing models](https://wilmott.com/automatic-for-the-greeks/) can be differentiated with ease.
```

Comment:
```suggestion
A key advantage of Julia for this task is that it can be used to implement fundamental numerical libraries, like [differential equations solvers](http://juliadiffeq.org/) or [optimisation libraries](https://github.com/JuliaOpt/JuMP.jl); this neatly solves a growing need in the ML community, in which researchers backpropagate through high-performance code such as [ray tracers](https://people.csail.mit.edu/tzumao/diffrt/) and [physics engines](https://arxiv.org/abs/1611.01652), but the gradients must still be implemented by hand in C++. In contrast, since Julia’s implementations are written in Julia, everything from [ODEs](https://github.com/FluxML/model-zoo/blob/a243e8b192236c30064fcdb7a36f17f3b6823c34/other/diffeq/diffeq.jl) to [financial pricing models](https://wilmott.com/automatic-for-the-greeks/) can be differentiated with ease.
```

Reply:
Ha, well, I have to believe that "financial princing" is a thing.